### ISSUE TITLE
heredocs

### ISSUE THREAD

**Issue by anntzer** [zsh-users/zsh-syntax-highlighting#275]
Probably a rare case, but for the sake of completeness:
```
$ cat <<EOF >/tmp/foo
heredoc> foo
heredoc> EOF
```
The second and third lines are incorrectly highlighted as errors.

**Comment by danielshahaf** [comment]
Indeed.  How can that case be detected?  `$CONTEXT` doesn't have a different value here, and using `[[ $(print -rP %_) == … ]]` in z-sy-h's code expands `%_` according to the value it would have in the middle of that condition, rather than according to $BUFFER.  There may be another way; I only spent five minutes on this issue so far.

By the way, pressing <kbd>C-d</kbd> at the `heredoc> ` prompt completes commands, which is arguably an upstream bug...

**Comment by danielshahaf** [comment]
How _should_ the heredoc be highlighted?  I can see the argument for applying `''` or `""` highlighting depending on whether `$foo` parameter references will be interpolated, but if we can't determine whether or not that is the case, applying no highlighting at all might be better.

**Comment by danielshahaf** [comment]
> Indeed.  How can that case be detected?  `$CONTEXT` doesn't have a different value here, and using `[[ $(print -rP %_) == … ]]` in z-sy-h's code expands `%_` according to the value it would have in the middle of that condition, rather than according to $BUFFER.  There may be another way; I only spent five minutes on this issue so far.

Actually, the printed value is the concatenation of the value in BUFFER and the value in the widget's code (e.g., if BUFFER has `while` and the widget code is `if`, then the value would be `while if`), so it _might_ be possible to look for the value `heredoc` specifically, because the widget-code part won't (or at least, "is unlikely to") have an (unfinished) heredoc.

It'd be nice if zsh upstream provided a cleaner solution.

> By the way, pressing <kbd>C-d</kbd> at the `heredoc> ` prompt completes commands, which is arguably an upstream bug...

Haven't looked into this.

**Comment by danielshahaf** [comment]
I guess we could solve this the same way zsh's own parser does: by remembering each `<<foo` as we encounter it and unwinding the queue after the line ends.  (It'll have to be a queue because of things like `diff <<foo <<bar`.)
